src/shortestpaths/astar.jl

```diff
@@ -56,38 +48,37 @@ function a_star_impl!(g, # the graph
 end
 
 """
-    a_star(g, s, t[, distmx][, heuristic])
+    a_star(g, s, t[, distmx][, heuristic][, edgetype_to_return])
+
+Compute a shortest path using the [A* search algorithm](http://en.wikipedia.org/wiki/A%2A_search_algorithm).
 
-Return a vector of edges comprising the shortest path between vertices `s` and `t`
-using the [A* search algorithm](http://en.wikipedia.org/wiki/A%2A_search_algorithm).
-An optional heuristic function and edge distance matrix may be supplied. If missing,
-the distance matrix is set to [`Graphs.DefaultDistance`](@ref) and the heuristic is set to
-`n -> 0`.
+# Arguments
+- `g::AbstractGraph`: the graph
+- `s::Integer`: the source vertex
+- `t::Integer`: the target vertex
+- `distmx::AbstractMatrix`: an optional (possibly sparse) `n × n` matrix of edge weights. It is set to `weights(g)` by default (which itself falls back on [`Graphs.DefaultDistance`](@ref)).
+- `heuristic`: an optional callable mapping each vertex to a lower estimate of the remaining distance from `v` to `t`. It is set to `v -> 0` by default (which corresponds to Dijkstra's algorithm)
+- `edgetype_to_return::Type{E}`: the eltype `E<:AbstractEdge` of the vector of edges returned. It is set to `edgetype(g)` by default. Note that the two-argument constructor `E(u, v)` must be defined, even for weighted edges: if it isn't, consider using `E = Graphs.SimpleEdge`.
 """
 function a_star(g::AbstractGraph{U},  # the g
     s::Integer,                       # the start vertex
     t::Integer,                       # the end vertex
     distmx::AbstractMatrix{T}=weights(g),
-    heuristic::Function=n -> zero(T)) where {T, U}
-
-    E = Edge{eltype(g)}
-
+    heuristic=n -> zero(T),
+    edgetype_to_return::Type{E}=edgetype(g)) where {T, U, E<:AbstractEdge}
```

Comment:
Isn't this a bit dangerous with respect to the other optional argument `heuristic` ?

Reply:
I agree that it may be dangerous, see [here](https://github.com/JuliaGraphs/Graphs.jl/pull/125#discussion_r878798977) for why I did it. We just have to weigh the pros and cons

Reply:
I added it back for now to facilitate merging, we can re-discuss it another time